src/librustc_typeck/check/method/suggest.rs

```diff
@@ -518,7 +518,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     }
                 }
 
-                if let Some(lev_candidate) = lev_candidate {
+                fn is_str_ref(ty: Ty<'_>) -> bool {
+                    match ty.sty {
+                        ty::Str => true,
+                        ty::Ref(_, ty, _) => is_str_ref(&ty),
+                        _ => false,
+                    }
+                }
+                if item_name.as_str() == "as_str" && is_str_ref(&actual) {
+                    err.span_suggestion(
+                        span,
```

Comment:
Sadly this span is not correct, it should include the whole method call, including the parentheses, and this one only points at the method identifier. If we could fix that, then we could make this `Applicability::MachineApplicable`. I'll need to take a second look to see how hard it'll be to procure the parent expression to get the correct span. If it is too hard, then this could be ok (as is with `MaybeIncorrect`). Would you mind taking a look yourself to see if you can get the right span from somewhere up the chain?

Reply:
Sorry for late reply. I'm readig the corresponding source code, and I think if we don't need the accurate HIR info, we could just search through the source_map and find the prev "." and following "()" and then generate the new span. If you think this is acceptable, I'll implement this.

Reply:
I would prefer to leave a `FIXME` comment and change this to be a `span_label` instead of a `span_suggestion` in that case. We do that kind of span wrangling in a couple of places but it's very bug-prone.

Reply:
Sure. Added a `FIXME` and converted it to `span_label`.